tools/taskcluster-worker-runner/protocol/protocol.go

```diff
@@ -62,28 +63,31 @@ func listOfStrings(val interface{}) []string {
 func (prot *Protocol) Start(asWorker bool) {
 	if asWorker {
 		prot.Register("welcome", func(msg Message) {
-			prot.RemoteCapabilities = FromCapabilitiesList(listOfStrings(msg.Properties["capabilities"]))
+			prot.remoteCapabilities = FromCapabilitiesList(listOfStrings(msg.Properties["capabilities"]))
 			prot.Send(Message{
 				Type: "hello",
 				Properties: map[string]interface{}{
-					"capabilities": prot.Capabilities.List(),
+					"capabilities": prot.localCapabilities.List(),
 				},
 			})
 			prot.SetInitialized()
 		})
 	} else {
 		prot.Register("hello", func(msg Message) {
-			prot.RemoteCapabilities = FromCapabilitiesList(listOfStrings(msg.Properties["capabilities"]))
+			prot.remoteCapabilities = FromCapabilitiesList(listOfStrings(msg.Properties["capabilities"]))
 			prot.SetInitialized()
 		})
 
 		prot.Send(Message{
 			Type: "welcome",
 			Properties: map[string]interface{}{
-				"capabilities": prot.Capabilities.List(),
+				"capabilities": prot.localCapabilities.List(),
 			},
 		})
 	}
+	prot.startedMutex.Lock()
```

Comment:
You should lock at the beginning of the function. If not, the race condition persists, as you can add a new capability in the middle of the capability negotiation.

Reply:
The code above here just registers callbacks -- the start only occurs in start of the receiving goroutine in the last line (after the unlock).

Reply:
Ah, in this case, the `mutex` doesn't do much.

Reply:
It does make `go test -race` happy, at least :)